src/workerd/api/sql.h

```diff
@@ -107,6 +107,58 @@ class SqlStorage final: public jsg::Object, private SqliteDatabase::Regulator {
     kj::Maybe<kj::Array<jsg::JsRef<jsg::JsString>>> names;
   };
 
+  // A statement in the statement cache.
+  struct CachedStatement: public kj::Refcounted {
+    jsg::HashableV8Ref<v8::String> query;
+    size_t statementSize;
+    SqliteDatabase::Statement statement;
+    CachedColumnNames cachedColumnNames;
+    kj::ListLink<CachedStatement> lruLink;
+
+    CachedStatement(jsg::Lock& js,
+        SqlStorage& sqlStorage,
+        SqliteDatabase& db,
+        jsg::JsString jsQuery,
+        kj::String kjQuery)
+        : query(js.v8Isolate, jsQuery),
+          statementSize(kjQuery.size()),
+          statement(db.prepareMulti(sqlStorage, kj::mv(kjQuery))) {}
+  };
+
+  class StatementCacheCallbacks {
+  public:
+    inline const jsg::HashableV8Ref<v8::String>& keyForRow(
+        const kj::Rc<CachedStatement>& entry) const {
+      return entry->query;
+    }
+
+    inline bool matches(const kj::Rc<CachedStatement>& entry, jsg::JsString key) const {
+      return entry->query == key;
+    }
+    inline bool matches(
+        const kj::Rc<CachedStatement>& entry, const jsg::HashableV8Ref<v8::String>& key) const {
+      return entry->query == key;
+    }
+
+    inline auto hashCode(jsg::JsString key) const {
+      return key.hashCode();
+    }
+    inline auto hashCode(const jsg::HashableV8Ref<v8::String>& key) const {
+      return key.hashCode();
+    }
+  };
+
+  using StatementMap = kj::Table<kj::Rc<CachedStatement>, kj::HashIndex<StatementCacheCallbacks>>;
+
+  struct StatementCache {
+    StatementMap map;
+    kj::List<CachedStatement, &CachedStatement::lruLink> lru;
+    size_t totalSize = 0;
```

Comment:
Should totalSize maybe feed back to jsg memory tracking metrics somehow?  Intuitively, I'd guess one could maybe add it to the `jsgGetMemorySelfSize()` return value, but I'm not familiar with exactly how that gets used.

Reply:
Hmm I looked at `SqlStorage::visitForMemoryInfo` and I frankly have no idea what's going on there. It seems like it's double-counting pointer sizes but not counting the things they point to. Also I suppose it cannot dereference `IoOwn` because `visitForMemoryInfo()` is not necessarily called from within any particular `IoContext`? So is it even possible to count the statement cache?

I think I'm just not going to touch it.